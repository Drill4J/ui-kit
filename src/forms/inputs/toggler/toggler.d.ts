import { TogglerProps } from './toggler-types';

export declare const Toggler: {
  ({
    className, label, value, onChange,
  }: TogglerProps): JSX.Element;
  displayName: string;
  __docgenInfo: {
    description: string;
    displayName: string;
    props: {
      className: {
        defaultValue: null;
        description: string;
        name: string;
        required: boolean;
        type: {
          name: string;
        };
      };
      label: {
        defaultValue: null;
        description: string;
        name: string;
        required: boolean;
        type: {
          name: string;
        };
      };
      value: {
        defaultValue: null;
        description: string;
        name: string;
        required: boolean;
        type: {
          name: string;
        };
      };
      disabled: {
        defaultValue: null;
        description: string;
        name: string;
        required: boolean;
        type: {
          name: string;
        };
      };
      onChange: {
        defaultValue: {
          value: string;
        };
        description: string;
        name: string;
        required: boolean;
        type: {
          name: string;
        };
      };
    };
  };
};
