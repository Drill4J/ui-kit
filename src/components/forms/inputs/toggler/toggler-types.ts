export interface TogglerProps {
  label?: React.ReactNode;
  value?: any;
  disabled?: boolean;
  onChange?: () => void;
}
