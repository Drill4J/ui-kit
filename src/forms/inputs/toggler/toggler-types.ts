export interface TogglerProps {
  className?: string;
  label?: React.ReactNode;
  value?: string | boolean;
  disabled?: boolean;
  size?: 'small';
  onChange?: () => void;
}
