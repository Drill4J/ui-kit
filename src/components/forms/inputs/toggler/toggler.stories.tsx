import React, { useState } from 'react';
import { ComponentMeta } from '@storybook/react';
import 'twin.macro';

import { Toggler } from './toggler';

export default {
  title: 'Toggler',
  component: Toggler,
} as ComponentMeta<typeof Toggler>;

const Template = (args) => {
  const { value, label, disabled } = args;
  const [val, setValue] = useState(value);
  return (
    <Toggler
      value={val}
      label={label}
      disabled={disabled}
      onChange={() => (val ? setValue(false) : setValue(true))}
    />
  );
};

export const Default = Template.bind({});
Default.args = {
  value: true,
  label: 'toggler',
  disabled: false,
};

export const Disabled = Template.bind({});
Disabled.args = {
  value: true,
  label: 'toggler',
  disabled: true,
};
