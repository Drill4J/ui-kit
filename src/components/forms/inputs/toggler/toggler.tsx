import tw, { styled, css } from 'twin.macro';

import { Input } from '../input';
import { TogglerProps } from './toggler-types';

export const Toggler = ({
  label, value, onChange = () => {}, disabled,
}: TogglerProps) => (
  <Checkbox disabled={disabled}>
    <CheckboxInput type="checkbox" checked={Boolean(value)} value={value} onChange={onChange} />
    <CheckboxTogglerLabel checked={Boolean(value)} />
    {label && <Label checked={Boolean(value)}>{label}</Label>}
  </Checkbox>
);

const Checkbox = styled.label<{disabled?: boolean}>`
  position: relative;
  display: inline-block;
  width: 20px;
  height: 12px;
  ${({ disabled }) => disabled && css`
    opacity: 0.3;
    pointer-events: none;
    cursor: default;
  `}
`;

const CheckboxInput = styled(Input)`
  opacity: 0;
  width: 0;
  height: 0;

  input {
    opacity: 0;
    width: 0;
    height: 0;
    margin: 0;
  }
`;

const CheckboxTogglerLabel = styled.span<{checked?: boolean}>`
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  ${tw`bg-monochrome-default`};
  transition: 0.4s;
  border-radius: 6px;

  &::before {
    position: absolute;
    content: '';
    height: 10px;
    width: 10px;
    left: 1px;
    bottom: 1px;
    ${tw`bg-monochrome-white`};
    transition: 0.4s;
    border-radius: 50%;
  }
  ${({ checked }) => checked && css`
    background-color: #00b602;

    &:hover {
      background-color: #33c535;
    }
    
    &:before {
      transform: translateX(8px);
    }
  `}
`;

const Label = styled.div<{checked?: boolean }>`
  position: absolute;
  left: 28px;
  top: -4px;
  font-size: 14px;
  color: #687481;
  white-space: nowrap;
  ${({ checked }) => checked && css`color: #00b602`}
`;
