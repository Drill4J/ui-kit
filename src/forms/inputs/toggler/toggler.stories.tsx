import React from 'react';
import { storiesOf } from '@storybook/react';

import { Toggler } from './toggler';

storiesOf('Toggler', module).add('Toggler', () => {
  const [value, setValue] = React.useState(true);
  return (
    <div style={{ display: 'flex', justifyContent: 'space-between', width: '300px' }}>
      <Toggler
        label="toggler"
        value={value}
        onChange={() => (value ? setValue(false) : setValue(true))}
      />
      <Toggler
        label="toggler"
        value={value}
        onChange={() => (value ? setValue(false) : setValue(true))}
        disabled
      />
    </div>
  );
});
