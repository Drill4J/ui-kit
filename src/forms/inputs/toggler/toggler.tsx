import { BEM } from '@redneckz/react-bem-helper';

import { Input } from '../input';

import { TogglerProps } from './toggler-types';

import styles from './toggler.module.scss';

const checkbox = BEM(styles);

export const Toggler = checkbox(
  ({
    className, label, value, onChange = () => {},
  }: TogglerProps) => (
    <label className={className}>
      <CheckboxInput type="checkbox" checked={Boolean(value)} value={value} onChange={onChange} />
      <CheckboxTogglerLabel />
      {label && <Label checked={Boolean(value)}>{label}</Label>}
    </label>
  ),
);

const CheckboxInput = checkbox.input(Input);
const CheckboxTogglerLabel = checkbox.slider('span');
const Label = checkbox.label('div');
