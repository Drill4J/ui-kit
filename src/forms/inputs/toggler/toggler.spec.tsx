import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { Toggler } from './toggler';

describe('Toggler', () => {
  it('Should render a toggler', () => {
    const tree = renderer
      .create(
        <Toggler
          label="toggler"
          onChange={() => {}}
          value
        />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should toggle the state', () => {
    const mockFn = jest.fn();
    const props = {
      label: 'toggler',
      onChange: mockFn,
      value: false,
    };
    const wrapper = mount(
      <Toggler {...props} />,
    );
    wrapper.find('input').simulate('change');
    expect(mockFn).toHaveBeenCalled();
  });
});
