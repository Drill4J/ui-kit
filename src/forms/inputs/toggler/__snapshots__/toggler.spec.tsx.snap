// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggler Should render a toggler 1`] = `
<label
  className="sc-gsDJrp dbMMPX"
>
  <div
    className="sc-dkPtyc cpUwga"
  >
    <input
      checked={true}
      className="sc-bdvvaa bVyvoG"
      onChange={[Function]}
      type="checkbox"
      value={true}
    />
  </div>
  <span
    checked={true}
    className="sc-hKwCoD fnERlt"
  />
  <div
    checked={true}
    className="sc-eCImvq bwvyzH"
  >
    toggler
  </div>
</label>
`;
